compiler/src/iree/compiler/Codegen/LLVMCPU/test/hal_interface_bindings.mlir

```diff
@@ -1,27 +1,87 @@
+  %memref = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%c72) : memref<?x2xf32, strided<[2, 1], offset: 18>>{%c128}
 
   // CHECK: %[[OFFSET_D0:.+]] = llvm.mul %[[C5]], %[[C2]]
-  // CHECK: %[[OFFSET_D1:.+]] = llvm.add %[[OFFSET_D0]], %[[C1]]
-  // CHECK: %[[OFFSET_PTR:.+]] = llvm.getelementptr %[[BUFFER]][%[[OFFSET_D1]]]
+  // CHECK: %[[INDEX0:.+]] = llvm.add %[[OFFSET_D0]], %[[C18]]
+  // CHECK: %[[INDEX1:.+]] = llvm.add %[[INDEX0]], %[[C1]]
+  // CHECK: %[[OFFSET_PTR:.+]] = llvm.getelementptr %[[BASE_PTR]][%[[INDEX1]]]
   // CHECK: %[[VALUE:.+]] = llvm.load %[[OFFSET_PTR]]
   %c1 = arith.constant 1 : index
   %c5 = arith.constant 5 : index
-  %value = memref.load %memref[%c5, %c1] : memref<?x2xf32>
+  %value = memref.load %memref[%c5, %c1] : memref<?x2xf32, strided<[2, 1], offset: 18>>
+
+  // CHECK: llvm.call @sink(%[[VALUE]])
+  llvm.call @sink(%value) : (f32) -> ()
+  return
+}
+llvm.func @sink(%arg0: f32) {
+  llvm.return
+}
+
+// -----
+
+// CHECK-LABEL: llvm.func @binding_ptrs_dynamic(
+func.func @binding_ptrs_dynamic() {
+  // CHECK-DAG: %[[C1:.+]] = llvm.mlir.constant(1
+  // CHECK-DAG: %[[C4:.+]] = llvm.mlir.constant(4
+  // CHECK-DAG: %[[C7:.+]] = llvm.mlir.constant(7
+  // CHECK-DAG: %[[C5:.+]] = llvm.mlir.constant(5
+  // CHECK-DAG: %[[C3:.+]] = llvm.mlir.constant(3
+
+  // CHECK: %[[STATE:.+]] = llvm.load %arg1
+  // CHECK: %[[CONSTANT_BASEPTR:.+]] = llvm.extractvalue %[[STATE]][9]
+  // CHECK: %[[OFFSET:.+]] = llvm.load %[[CONSTANT_BASEPTR]]
+  // CHECK: %[[OFFSET_ZEXT:.+]] = llvm.zext %[[OFFSET]]
+  %offset = hal.interface.constant.load[0] : index
+
+  // CHECK: %[[STATE0:.+]] = llvm.load %arg1
+  // CHECK: %[[CONSTANT_BASEPTR:.+]] = llvm.extractvalue %[[STATE0]][9]
+  // CHECK: %[[DIM1_PTR:.+]] = llvm.getelementptr %[[CONSTANT_BASEPTR]][2]
+  // CHECK: %[[DIM1:.+]] = llvm.load %[[DIM1_PTR]]
+  // CHECK: %[[DIM1_ZEXT:.+]] = llvm.zext %[[DIM1]]
+  // CHECK: %[[STATE1:.+]] = llvm.load %arg1
+  // CHECK: %[[CONSTANT_BASEPTR0:.+]] = llvm.extractvalue %[[STATE1]][9]
+  // CHECK: %[[DIM2_PTR:.+]] = llvm.getelementptr %[[CONSTANT_BASEPTR0]][3]
+  // CHECK: %[[DIM2:.+]] = llvm.load %[[DIM2_PTR]]
+  // CHECK: %[[DIM2_ZEXT:.+]] = llvm.zext %[[DIM2]]
+  %dim0 = hal.interface.constant.load[1]: index
+  %dim1 = hal.interface.constant.load[2] : index
+  %dim2 = hal.interface.constant.load[3] : index
+
+  // CHECK: %[[STATE3:.+]] = llvm.load %arg1
+  // CHECK: %[[BINDING_PTRS:.+]] = llvm.extractvalue %[[STATE3]][10]
+  // CHECK: %[[ARRAY_PTR:.+]] = llvm.getelementptr %[[BINDING_PTRS]][1] : (!llvm.ptr) -> !llvm.ptr, !llvm.ptr
+  // CHECK: %[[BASE_PTR:.+]] = llvm.load %[[ARRAY_PTR]] : !llvm.ptr -> !llvm.ptr
+  %memref = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) offset(%offset) : memref<?x?x?xf32, strided<[?, ?, 1], offset: ?>>{%dim0, %dim1, %dim2}
+
+  // CHECK: %[[BASE_OFFSET:.+]] = llvm.udiv %[[OFFSET_ZEXT]], %[[C4]]
```

Comment:
This looks fishy and lossy, would be better to add the byte offset to the ptr and then compute the address of the element.

Reply:
Actually on second thought, I think it is better to leave this as is. The need here is to build the memref descriptor object from the given ABI information. If the offset is constant you get that built using just the type. If the offset is not constant you set the offset using the byte offset provided in the `hal.interface.binding.subspan` and converting it to element offset. So here we are converting from the byte offset to element offset by dividing by 4. Doing a GEP instead would mean
- For constant offset we will get the memref descriptor to get the offset
- For the dynamic offset with will do GEP to move the base pointer and then set the offset to be 0. I think that is confusing. Would be better to keep the dynamic and static behavior consistent.

So what is the lossy part of doing the division here?

Reply:
Ok, how about footgun then ?

I believe the current calculation is wrong.
Example: offset = 18, type = f32

Address of 3rd f32 from the base should be 18 + 3 * 4 = 30.
Address that I think is currently calculated (18/4 + 3) * 4 = 28.

This goes back to my comment about reusable APIs for non-trivial things: we need 1 single place for all these things upstream.

Reply:
> Ok, how about footgun then ?

The point was to not have snark in reviews, and avoid being dismissive. It is a request, so its your choice to accept/reject the request.

> 
> I believe the current calculation is wrong. Example: offset = 18, type = f32
> 
> Address of 3rd f32 from the base should be 18 + 3 * 4 = 30. Address that I think is currently calculated (18/4 + 3) * 4 = 28.
> 
> This goes back to my comment about reusable APIs for non-trivial things: we need 1 single place for all these things upstream.

I am not sure this is correct. offset of 18 for type of f32 is illdefined. The first element is not aligned, and that would be an illdefined program (we can catch that statically but cant catch that dynamically).
So lets take an aligned offset of 20 bytes. then the third element position could be computed as (20/4 + 3) words or (20 / 4 + 3) * 4 bytes. I dont see any loss here. IIUC in a static world this would be represented as `memref<?xf32, strided<[], offset : 5>>` (assuming my understanding is correct that offsets in `memref` are based on number of elements and not number of bytes). So here to the base pointer of the memref is not expected to be misaligned. It has to aligned to word boundary.

Reply:
This is correct for stuff coming from a memref descriptor.
In this case, it is coming from the function arguments and I am unclear that IREE abides by that convention.

The subspan op says:
```
    An optional alignment indicates the byte alignment of the base binding
    resource. Note that the byte offset is added to the base and the alignment
    will be the minimum of the two.
```

Edit: sync'ed offline with @MaheshRavishankar and apparently the offset has to be a multiple of the element type, alleviating my concerns here.

Reply:
Could you write that down in the comments?
Eventually that would be nice to codegen some sort of runtime asserts (under a flag) to help debugging this kind of things if something ever breaks that.